subsys/net/ip/tcp.c

```diff
@@ -2639,117 +2639,138 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
+			 * FIN & -   -> TCP_CLOSING
+			 * -   & ACK -> TCP_FIN_WAIT_2
+			 * FIN & ACK -> TCP_TIME_WAIT
 			 */
-			if (FL(&fl, ==, (FIN | ACK),
-			    (th_ack(th) == conn->seq) && (th_seq(th) == conn->ack))) {
-				tcp_send_timer_cancel(conn);
-				conn_ack(conn, + 1);
-				tcp_out(conn, ACK);
-				next = TCP_TIME_WAIT;
-				NET_DBG("FIN acknowledged, going to TIME_WAIT_STATE");
-				verdict = NET_OK;
-			} else if (th &&
-				   (FL(&fl, ==, FIN, th_seq(th) == conn->ack) ||
-				    FL(&fl, ==, (FIN | ACK), th_seq(th) == conn->ack))) {
-				tcp_send_timer_cancel(conn);
-				conn_ack(conn, + 1);
-				NET_DBG("FIN not yet acknowleged, going to CLOSING state");
-				if (th_ack(th) != conn->seq) {
-					/* Not acknowledged the FIN flag yet, so resend */
-					tcp_out_ext(conn, (FIN | ACK), NULL, conn->seq - 1);
-				} else {
-					tcp_out(conn, ACK);
-				}
-				next = TCP_CLOSING;
-				verdict = NET_OK;
-			} else if (th && FL(&fl, ==, ACK, th_ack(th) == conn->seq)) {
-				tcp_send_timer_cancel(conn);
-				next = TCP_FIN_WAIT_2;
-				verdict = NET_OK;
-				NET_DBG("FIN acknowledged, going to FIN_WAIT_2 state seq %u, ack %u"
-					, conn->seq, conn->ack);
 
-				/* Acknowledge any received data */
-				if (len > 0) {
-					tcp_out(conn, ACK);
+			/*
+			 * FIN2:
+			 * Only FIN is relevant in this state, as FIN has already acknowledged
+			 * -   -> TCP_FIN_WAIT_2
+			 * FIN -> TCP_TIME_WAIT
+			 */
+
+			/*
+			 * Closing:
+			 * Our FIN has to be acknowledged
+			 * -   -> TCP_CLOSING
+			 * ACK -> TCP_TIME_WAIT
+			 */
+			if (FL(&fl, &, ACK, th_ack(th) == conn->seq)) {
+				if (conn->state == TCP_FIN_WAIT_1) {
+					NET_DBG("FIN acknowledged, going to FIN_WAIT_2 "
+						"state seq %u, ack %u"
+						, conn->seq, conn->ack);
+					tcp_send_timer_cancel(conn);
+					next = TCP_FIN_WAIT_2;
 				}
-			} else if (th) {
-				NET_DBG("Staying in FIN_WAIT_1 flags 0x%x state seq %u, ack %u",
-					fl, conn->seq, conn->ack);
-				if (len > 0) {
+				if (conn->state == TCP_CLOSING) {
+					NET_DBG("FIN acknowledged, going to TIME WAIT "
+						"state seq %u, ack %u"
+						, conn->seq, conn->ack);
 					tcp_send_timer_cancel(conn);
-					/* Send out a duplicate ACK, with the pending FIN flag */
-					tcp_out(conn, FIN | ACK);
+					next = TCP_TIME_WAIT;
 				}
 				verdict = NET_OK;
 			}
-		}
-		break;
-	case TCP_FIN_WAIT_2:
-		/* Acknowledge but drop any data, but subtract any duplicate data */
-		if (th) {
-			if (len > 0) {
-				int32_t new_len = len - net_tcp_seq_cmp(conn->ack, th_seq(th));
+			if ((next == TCP_FIN_WAIT_2) ||
+			    (conn->state == TCP_FIN_WAIT_2) ||
+			    (conn->state == TCP_CLOSING)) {
 				/* Cases:
 				 * - Data already received earlier: len > 0 , new_len <= 0
 				 * - Partially new data len > 0, new_len > 0
 				 * - Out of order data len > 0, new_len > 0, ignore the data in
 				 *   between
 				 */
 				if (new_len > 0) {
-					conn_ack(conn, + new_len);
+					/* This should not happen, since we do not implement half
```

Comment:
Is `This should not happen` correct here? Sure, we do not support half-closed streams, but `This should not happen` indicates that the peer did something against the standard.

Reply:
True, only in the CLOSING state it should not happen, in FIN1 and FIN2 it is allowed. Modified the comment.